utils/src/lib.rs

```diff
@@ -72,6 +72,65 @@ pub fn hash_to_bn254_field_size_be(bytes: &[u8]) -> Option<([u8; 32], u8)> {
     None
 }
 
+/// Hashes the provided `bytes` with Keccak256 and ensures the result fits
+/// in the BN254 prime field by repeatedly hashing the inputs with various
+/// "bump seeds" and truncating the resulting hash to 31 bytes.
+///
+/// The attempted "bump seeds" are bytes from 255 to 0.
+///
+/// # Examples
+///
+/// ```
+/// use light_utils::hashv_to_bn254_field_size_be;
+///
+/// hashv_to_bn254_field_size_be(&[b"foo", b"bar"]);
+/// ```
+pub fn hashv_to_bn254_field_size_be(bytes: &[&[u8]]) -> ([u8; 32], u8) {
+    // Loops with decreasing bump seed to find a valid hash which is less than
+    // bn254 Fr modulo field size.
+    for bump_seed in (0..=u8::MAX).rev() {
+        // NOTE(vadorovsky): This is, to my current knowledge, the least bad
+        // solution for the problem of "adding one element to a slice".
+        //
+        // The `bytes` slice contains another slices/references to the actual
+        // bytes. Therefore, adding these slices to the new vector doesn't
+        // create a copy of the underlying bytes. What's more, adding the
+        // slices to the `inputs` vector bints it to the lifefime of `bytes`.
+        //
+        // However, even though copies of the underlying bytes are not being
+        // made, the copies of slices, as "fat pointers", are. Each slice takes
+        // 16 bytes. Therefore, each iteration allocate aproximately
+        // 16 * (n + 1) bytes. In the most cases, this loop should not iterate
+        // more than once, so it's safe to assume that this function will
+        // allocate 16 * (n + 1) all together, or twice this value in the worst
+        // case. Due to the nature of Solana bump allocator, that memory is not
+        // going to be freed after going out of this function's scope, if the
+        // function is used on-chain. If that's problematic, the
+        // `#[heap_neutral]` macro can be used as a workaround.
+        let mut inputs = Vec::with_capacity(bytes.len() + 1);
+        inputs.extend(bytes);
+        let bump_seed = [bump_seed];
+        inputs.push(bump_seed.as_slice());
+        {
```

Comment:
We can probably omit the bump seed all together since we are truncating to 248 bytes there is no way for the number to be greater than field size.

Reply:
Cool. Was there any particular reason behind introducing the loop in your helper? Or do you consider in unnecessary there as well?

For me the lack of necessity for it makes sense, any 31-byte hash will fit in Fr.

Reply:
The loop is a remainder from the initial design which didn't truncate to 31 bytes and just searches for a bump seed which yields a hash less than field size. However, in some cases tens of hashes are required which is too expensive onchain.